flowd/flowd.1
.\" @(#) $Header: /usr/local/cvsroot/netflow/netflow/flowd/flowd.1,v 1.1.1.1 2004/05/29 09:06:41 harry Exp $
.\"
.TH FLOWD 1  "13 Mar 2001"
.SH NAME
flowd \- dump Cisco netflow data to files
.SH SYNOPSIS
.na
.B flowd
[
.B \-chxn
] [
.B \-t
.I hours
] [
.B \-f
.I flush
] [
.B \-r
.I ratelimit
] 
.br
.ti +8
[
.B \-p
.I port
] [
.B \-d
.I dumpdir
]
.br
.ad
.SH DESCRIPTION
.LP
\fIflowd\fP dumps flows from a Cisco netflow feed to a cyclic series of files
.I flowd.
.LP
Run
.I flowd
as root if 
.I \-c 
is given since it will need to attach to the BGP shared memory segment
created by the bgp process.
.SH OPTIONS
.TP
.B \-h
Help. Show command usage
.TP
.B \-x
Don't fork and run as foreground process. The default is to run as daemon
.TP
.B \-n
Drop flows going to interface index 0.  These are flows of packets that have been
dropped by the router due to its rate limiting policy. Usually as a result of a DoS
This will also prevent huge trace files.
.TP
.B \-p port
The UDP port to listen on for netflow packets. The default is 9992
.TP
.B \-c
Dump BGP communities information also (4 max). This increases the data per flow
from 46 bytes to 64 bytes.
.TP
.B \-t Hours
The number of hours of history to keep. Data files will be recycled within this period.
.TP
.B \-f Flush
The number of seconds to accumulate flows in memory, after which time these are flushed
to the next data file. One good reason to flush data often, is that memory consumption
can be an issue. You don't want this to get into a swap situation as flow collection
may suffer.
.TP
.B \-d Dumpdir
The directory where the data files are to be written. The default is /var/log/netflow/trace
.TP
.B \-r rate
Rate limiting is provided to give some protection against DOS attacks and the make the
maximum file size somewhat predictable. (64 bytes per flow with BGP communities,
and 46 bytes without). When the rate limit is breached, netflow packets are dropped.
This is based on a per second bases.  So within the second some but not all data will
be lost. 
.br

To keep 24Hrs of data at a sustained flowrate of 10,000 flows/sec at 64 byte flow
records would require approximately 55 Gbytes.
.br

Identical flows are combined by adding the octets to and from and incrementing 
the flowcount indicator (See output Format). Flows are also tested for bi-directionality
and combined into one. This attempts to save space.

A log 
.I flowd.log
is written to the data directory containing various diagnostics. 

.SH OUTPUT FORMAT

The output is a set of binary data with the following C-structures.
.br

Firstly a small header is written
.br
.nf

typedef struct
{
    char  sig1;                    /* File signature 'F' */
    char  sig2;                    /* File signature 'O' */
    char  sig3;                    /* File signature 'O' */
    unsigned char  version;        /* 10 = 1.0, 44=4.4,  200=20.0  */
    char  options[4];              /* options affecting size of struct */
} header;

This is followed by the flowdata structures as:

typedef struct
{
    time_t    timestamp;   /* From unix_secs in the netflow header */
    ulong_t   srcIP;       /* Source IP address */
    ulong_t   dstIP;       /* Destination IP address */
    ulong_t   packets_to;  /* Number of packets in the flow */
    ulong_t   packets_from;/* Number of packets in the flow */
    ulong_t   bytes_to;    /* Number of bytes in the flow */
    ulong_t   bytes_from;  /* Number of bytes in the flow */
    ushort_t  srcIDB;      /* Source interface */
    ushort_t  dstIDB;      /* Destination interface */
    ushort_t  srcport;     /* Source port */
    ushort_t  dstport;     /* Destination port */
    ushort_t  protocol;    /* Protocol number */
    ushort_t  srcAS;       /* Peer AS of source address */
    ushort_t  dstAS;       /* Peer AS of destination address */
    ushort_t  bi;          /* 1=bi-directional , 0=Uni-directional */
    ushort_t  flows;       /* Number of flows for this one */
                           /* size = 46 */
    ushort_t  communities_len;
    uint32_t  community1;   
    uint32_t  community2;   
    uint32_t  community3;   
    uint32_t  community4;   
                           /* size=64 */
} tracedetails;

If no BGP communities are being collected, the first 46 bytes only are used.
.br

.fi

.SH SIGNALS
The 
.I SIGUSR1 
signal will immediatly flush to the next trace file.
The 
.I SIGUSR2 
signal toggles debug on/off. This will create a file "debug.0"
that is similar to the trace files but outputs the flows without tallying
duplicates and reverse flows.

.SH EXAMPLES
.LP
To store 24 hours of history with data being saved every 5 minutes with BGP
communities.
.RS
.br

.nf
\fBflowd -t 24 -f 300 -c \fP
.fi
.RE
.SH "SEE ALSO"
.I flowread(1)
.SH AUTHORS
Harry Raaymakers
.br
Jim Crumpler ,bgp API and ip2cidr lib
.br
Eric Sven Ristad and Peter N. Yianilos, libpa library

.br
.SH BUGS
Please send bug reports to harryr@connect.com.au


flowd/flowd.c
/***
 * 
 * Raw netflow collector 
 *
 * Options: See ./collect -h
 *
 * $Id: flowd.c,v 1.1.1.1 2004/05/29 09:06:41 harry Exp $
 * $Source: /usr/local/cvsroot/netflow/netflow/flowd/flowd.c,v $
 *
 ***/

#include <stdio.h>
#include <sys/types.h>
#include <signal.h>
#include <strings.h>

#include <symbol.hash.h>
#include <bgp_api.h>
#include <ip2cidr.h>

#ifndef FALSE
#define FALSE (0)
#endif
#ifndef TRUE
#define TRUE (1)
#endif
#define PORT 9992

#define LOGLEVEL 2	/* default loglevel. 0=Nothing, 2=Informational, 4=Verbose */

int debug_hist=0;
int max_debug=0;
int debug_count=0;
int debug_flows=FALSE;	/* unmolested flows */

int trace_hist=0;
int max_trace=0;
int trace_count=0;

symbol_hash_t*   ip_hash=NULL;
int main_doflush = FALSE;
int zeroint=FALSE;

int flows_in_packet=0;
int rate_limit=999999;		/* huge default, rate limmiting, flows per second  */
int loglevel=LOGLEVEL;

static int flush_interval = 15 * 60;
char *trace_d = "/var/log/netflow/trace";
char *logfile = "/var/log/netflow/trace/flowd.log";
char *trace_dir = NULL;
const char *debug_dir = "/var/log/netflow/trace";
int		bgp_communities=FALSE;
int		interface_indexes=FALSE, sequence=FALSE;
char		options[4]={0,0,0,0};

void            usage(void);
void 		set_alarm();
static void     handle_usrsig(int sig, siginfo_t* sip, void* uap);
static void     signal_init(void);
char		*myname;

main(int argc, char *argv[])
{
    int c,i=0, limitting=FALSE, port=PORT;
    int total_flows=0;
    int nflows;			/* number of flows in a packet */
    extern char *optarg;
    time_t	t,prevt=time(NULL);
    pid_t 	pid, pgrp;
    int 	forkme=TRUE; 

    myname=argv[0];

    while ((c = getopt(argc, argv, "d:f:r:t:l:p:nchxsz")) != EOF)
    {
        switch (c) {
        case 'x':
            forkme=FALSE;
	    break;
        case 'n':			/* throw away flows to iface 0 */
	    zeroint=TRUE;
	    break;
        case 'c':			/* capture BGP communities */
            bgp_communities=TRUE;
	    options[i++]='c';
            break;
        case 'd':			/* storeage dir */
	    trace_dir=strdup(optarg);
            break;
        case 'f':			/* flush data every n-seconds */
            flush_interval = atoi(optarg);
            break;
        case 'h':			
		usage();
        case 'l':			/* logging level */
	    loglevel=atoi(optarg);
            break;
        case 'p':			/* UDP port to listen on */
            port = atoi(optarg);
            break;
        case 's':			/* sequence number check only */
            sequence = TRUE;
	    forkme=FALSE;
	    logfile= "/dev/null";
            break;
        case 'z':			/* No logfile */
	    logfile= "/dev/null";
	    break;
        case 'r':			/* Set rate limitting (DOS protection) */
            rate_limit = atoi(optarg);
            break;
        case 't':			/* Hours to keep */
            if (! (trace_hist = atoi(optarg) * 60 * 60))
                usage();
            max_trace=trace_hist / flush_interval;
            break;
	default:
		usage();
	}
    }

/* Daemon init */

    if (forkme==TRUE) {
        chdir("/");

        if ((pid=fork())>0) 
	    exit(0);
        else if (pid == -1) {
	    fprintf(stderr,"Failed to fork child\n");
	    exit(-1);
        }

        if ((pgrp = setsid()) == -1) {
	    fprintf(stderr,"Failed to become group leader\n");
	    exit(-1);
        }

        if (freopen("/dev/null","r",stdin) == NULL) {
	    fprintf(stderr,"Failed to reopen stdin to devnull\n");
	    exit(-1);
        }
        if (freopen("/dev/null","w",stdout) == NULL) {
	    fprintf(stderr,"Failed to reopen stdout to devnull\n");
	    exit(-1);
        }
    }

    if (!trace_dir)
	trace_dir=trace_d;

    init_logger(logfile);


    if (!sock_init(port)) {
	logger(1,"Failed to initialise socket on port: %d\n",port);
	exit(-1);
    }

    max_trace=trace_hist / flush_interval;

    signal_init();

    if (bgp_communities) { 
        if (bgp_db_attach(argv[0])!=1) {
	    logger(1,"Failed to attach to existing bgp shared mem segment.\n");
	    exit(-1);
	}
    }

    set_alarm();

    logger(2,"Starting Netflow collector ...\n");
    logger(2,"Rate limit=%d flows/sec History=%d seconds, %sflush interval=%d seconds. Produces %d files\n",rate_limit,trace_hist,(zeroint?"Dropping egress iface 0 flows, ":" "), flush_interval, max_trace);

    for(;;) {
        t=time(NULL);
	if (t>prevt) {
	    if (limitting==TRUE)
	        logger(2,"Rate is >=%d flows/sec. Dropping flows.\n",total_flows);
	    prevt=t;
	    logger(4,"nflows=%d\n",total_flows);
	    total_flows=0;
	    flows_in_packet=0;
	}

	if (total_flows < rate_limit) {
	    sock_handle_one_packet(sequence);
	    total_flows+=flows_in_packet;	
	    limitting=FALSE;
	} else
	    limitting=TRUE;

	if (main_doflush && !sequence) {
	    logger(3,"Signal recieved. Flushing data\n");
	    main_doflush=FALSE;
	    dump_results();
	}
    }
}

static void     handle_sigalarm(int sig, siginfo_t* sip, void* uap) {
    main_doflush++;
    set_alarm();
}

void
set_alarm()
{
    time_t              cur_time;
    time_t              next_flush;
    
    cur_time = time(NULL);
    next_flush = ((cur_time / flush_interval) + 1) * flush_interval;
    
    alarm(next_flush - cur_time);
}

static void     handle_usrsig(int sig, siginfo_t* sip, void* uap) {
    
    switch(sig) {
        
    case SIGUSR1: 		/* flush datafile */
        main_doflush = TRUE;
	logger(1,"SIGUSR1 recieved, flushing.\n");
        break;
        
    case SIGUSR2:		/* debug on/off */
        debug_flows = !debug_flows;
	logger(1,"SIGUSR2 recieved, debug is %d\n",debug_flows);
        break;
    }
}

/**
 * ALARM - flush results and rotate files 
 * HUP   - Re-read config file 
 ***/

static void signal_init(void) {

    struct sigaction     sa;
    
    sigemptyset(&sa.sa_mask);
    sa.sa_flags =  0;
    
    sigemptyset(&sa.sa_mask);

    sa.sa_handler = handle_sigalarm;
    sigaction(SIGALRM, &sa, NULL);

    sa.sa_handler = handle_usrsig;
    sigaction(SIGUSR1, &sa, NULL);
    sigaction(SIGUSR2, &sa, NULL);

}

void usage()
{
    printf("usage: %s [-n] [-h] [-t hours] [-f seconds] [-r flows] [-c]\n",myname);
    printf("\t-h = help\n");
    printf("\t-x = Don't fork.\n");
    printf("\t-t hours = Hours of history to keep before recycling data files.\n");
    printf("\t-f seconds = Flush interfal to dump data to next data file.\n");
    printf("\t-p port = UDP port to listen on for netflow data.\n");
    printf("\t-r flows = Rate limitting. Max number of flows per second before throwing away.\n");
    printf("\t-s Print NetFlow packet headers. Used to check sequence numbers. Wont fork.\n");
    printf("\t-c = Include upto 4 BGP communities on source address (increases data file size by 18 bytes per flow)\n");
    printf("\t-z =  No logfile\n");
    printf("\t-n = Throw away flows to interface 0. Router rate limmited.\n");
    exit(0);
}

